### PR TITLE
Frrist/update golang1.23.0

### PR DESCRIPTION
**Pull request by frrist**
## Why
In https://github.com/bacalhau-project/bacalhau/pull/4330/files#diff-77364c74880b2d28d60bdc250c7832430bb2c7d87d2e2359b05146c9f3cb5d6bR111 we are adding migration logic which requires directories to be copied. go 1.23 includes the [os.CopyFS](https://pkg.go.dev/os#CopyFS) method which is helpful for this task, and preferred to writing our own CopyFS implementation. 

## How
Developers will need to perform the following actions:
1.  [download and install go 1.23.0](https://go.dev/dl/)
3. [update golangci-lint](https://golangci-lint.run/welcome/install/#local-installation): `curl -sSfL https://raw.githubusercontent.com/golangci/golangci-lint/master/install.sh | sh -s -- -b $(go env GOPATH)/bin v1.60.1`

### CODE REVIEW

**Comment by frrist** [comment]
Updating to this version of golang requires we update the linter to match, it appears the buildkite instance lacks sufficient memory to run the liner now:
```
2024-08-16 10:39:44 PDT	golangci-lint............................................................Failed
2024-08-16 10:39:44 PDT	- hook id: golangci-lint
2024-08-16 10:39:44 PDT	- exit code: 137
2024-08-16 10:39:44 PDT	
2024-08-16 10:39:44 PDT	/buildkite/builds/4s62es3plflsi-1/expanso/bacalhau-golang/.gitprecommit/golangci-lint.sh: line 3:   450 Killed                  golangci-lint run --allow-parallel-runners --timeout 10m
2024-08-16 10:39:44 PDT	/buildkite/builds/4s62es3plflsi-1/expanso/bacalhau-golang/.gitprecommit/golangci-lint.sh: line 3:   452 Killed                  golangci-lint run --allow-parallel-runners --timeout 10m
2024-08-16 10:39:44 PDT	/buildkite/builds/4s62es3plflsi-1/expanso/bacalhau-golang/.gitprecommit/golangci-lint.sh: line 3:   453 Killed                  golangci-lint run --allow-parallel-runners --timeout 10m
2024-08-16 10:39:44 PDT	/buildkite/builds/4s62es3plflsi-1/expanso/bacalhau-golang/.gitprecommit/golangci-lint.sh: line 3:   454 Killed                  golangci-lint run --allow-parallel-runners --timeout 10m
```

File: .golangci.yml [hunk 1]
```diff
@@ -87,7 +81,6 @@ linters:
   - gocyclo
   - gofmt
   - goimports
-  - gomnd
```
Comment (wdbaruni) [inline .golangci.yml:90]: shouldn't we enable `mnd`?

File: ops/terraform/remote_files/scripts/install-node.sh [hunk 2]
```diff
@@ -8,7 +8,7 @@ source /terraform_node/variables
 function install-go() {
   echo "Installing Go..."
   rm -fr /usr/local/go /usr/local/bin/go
-  curl --silent --show-error --location --fail https://go.dev/dl/go1.21.8.linux-amd64.tar.gz | sudo tar --extract --gzip --file=- --directory=/usr/local
+  curl --silent --show-error --location --fail https://go.dev/dl/go1.23.0.linux-amd64.tar.gz | sudo tar --extract --gzip --file=- --directory=/usr/local
```
Comment (wdbaruni) [inline ops/terraform/remote_files/scripts/install-node.sh:11]: please deploy to staging to verify all is working as expected

**Review by wdbaruni — APPROVED** [review]

File: .golangci.yml (hunk 1 above)
Comment (frrist) [reply .golangci.yml:90]: `gomnd` [has been desecrated](https://github.com/golangci/golangci-lint/issues/4767#issuecomment-2135208130) - replaced by `mnd` which is declared here now: https://github.com/bacalhau-project/bacalhau/pull/4332/files#diff-6179837f7df53a6f05c522b6b7bb566d484d5465d9894fb04910dd08bb40dcc9R35 (line 35 of this PR)

Comment (wdbaruni) [reply .golangci.yml:90]: You still need to re-declare it here, right? Line 35 is the linter setting, but all linters are disabled except the ones listed here

File: ops/terraform/remote_files/scripts/install-node.sh (hunk 2 above)
Comment (frrist) [reply ops/terraform/remote_files/scripts/install-node.sh:11]: Done. working as intended on this commit https://github.com/bacalhau-project/bacalhau/pull/4332/commits/7404e375d09be15015438a2f10d6a17d18b83783:
```
export BACALHAU_API_HOST=bootstrap.staging.bacalhau.org
~> bacalhau agent version
Bacalhau v1.4.1-rc3-28-g7404e375d
BuildDate 2024-08-18 01:42:44 +0000 UTC
GitCommit 7404e375d09be15015438a2f10d6a17d18b83783

```

File: .golangci.yml (hunk 1 above)
Comment (frrist) [reply .golangci.yml:90]: I was initially confused as to why it ran without me setting it - I have modifed the pre-commit script to respect the linter config in the repo: https://github.com/bacalhau-project/bacalhau/pull/4332/commits/7404e375d09be15015438a2f10d6a17d18b83783

**Comment by frrist** [comment]
This PR is currently block on updating the docker container buildkit uses to build and test bacalhau. I have updated the Dockerfile for the image here: https://github.com/bacalhau-project/bacalhau/pull/4332/commits/3c9de726a4a73c009591534719a5fd9ba18f3427 but need direction from @udsamani on how to make this change take affect.

File: .golangci.yml [hunk 3]
```diff
@@ -156,14 +158,6 @@ run:
   # directory holds the correct copies of dependencies and ignores
   # the dependency descriptions in go.mod.
   #
-  # Allowed values: readonly|vendor|mod
-  # By default, it isn't set.
-  modules-download-mode: readonly
-  # Allow multiple parallel golangci-lint instances running.
-  # If false (default) - golangci-lint acquires file lock on start.
-  allow-parallel-runners: false
```
Comment (udsamani) [inline .golangci.yml:164]: Now that you have removed parallel runners from all places would it run serially or parallely ?

Comment (udsamani) [reply .golangci.yml:164]: IMO ! We need this https://github.com/golangci/golangci-lint/issues/869 ! This is the root cause failures after updating the golang image to 1.23.0

**Pull request merged**